Fix thread-safety issues

greenwald opened the pull request:
This addresses #182 and #205 

I have removed storage of the prior during calculation, since it can be reconstructed via (probability - likelihood). However, we don't even need to store it in the `TTree` as a branch. It could be replaced by an alias---which gets saved with the `TTree`. This has the advantage of lightening the footprint of the tree, but the disadvantage that the alias cannot be loaded into a variable the same way a branch can (via `SetBranchAddress`). It will make no difference for a user who draws directly from the `TTree`, but it will make a difference to someone who accesses it through a macro.

I have not been able to fix #205. But have found that the mismatch is not from the prior. It's actually from the likelihood. `LogProbability` agrees with parameter values stored in the tree; but `LogLikelihood` has occasional discrepancies. --> I am stuck.

fredRos commented:
It seems old values are written out. Initially likelihood and prior are set to -inf but we require that the chain never be in a state with -inf. Mismatches later may be due to same origin.
```
root [19] mismatch_mcmc->Scan("","LogLikelihood/LogLikelihood!=1")
************************************************************************************************
*    Row   * Chain.Cha * Iteration * Phase.Pha * LogProbab *         x * LogLikeli * LogPrior. *
************************************************************************************************
*        1 *         1 *         1 *        -1 * -2.569636 * 0.9533992 *      -inf *      -inf *
*        4 *         4 *         1 *        -1 * -1.762518 * 0.1244983 *      -inf *      -inf *
*        5 *         5 *         1 *        -1 * -0.012188 * 0.5312259 *      -inf *      -inf *
*        6 *         6 *         1 *        -1 * -0.001204 * 0.4901837 *      -inf *      -inf *
*       13 *         5 *         2 *        -1 * -0.012188 * 0.5312259 *      -inf *      -inf *
*       14 *         6 *         2 *        -1 * -0.001204 * 0.4901837 *      -inf *      -inf *
*       21 *         5 *         3 *        -1 * -0.012188 * 0.5312259 *      -inf *      -inf *
************************************************************************************************
```

greenwald commented:
I can remove the timing issue, by editing line 1529 of `BCEngineMCMC.cxx`:

If one changes

    fMCMCStates[chain].log_probability = p1;

where `p1 =  LogEval(fMCMCThreadLocalStorage[chain].parameters)`, to

    fMCMCStates[chain].log_probability = fMCMCThreadLocalStorage[chain].log_probability;

where the local storage's `log_probability` is set directly inside `LogProbabilityNN` (called by `LogEval`) at the same time as `log_likelihood` and `log_prior`.

HOWEVER, this then causes one of the tests to fail: one about output precision.

fredRos commented:
> The check failure seems also to stem from ROOT dictionary generation for structs declared inside BCEngineMCMC

So it fails with ROOT5. @oschulz is our expert on dictionary generation. Perhaps he knows how to massage ROOT5 the right way? ROOT6 apparently is clever enough.

```
libBAT_rdict.cxx: In function 'int G__libBAT_rdict_659_0_24(G__value*, const char*, G__param*, int)':
libBAT_rdict.cxx:6931:23: error: 'ChainState' was not declared in this scope
          const vector<ChainState>& obj = ((const BCEngineMCMC*) G__getstructoffset())->GetChainStates();
                       ^
libBAT_rdict.cxx:6931:33: error: template argument 1 is invalid
          const vector<ChainState>& obj = ((const BCEngineMCMC*) G__getstructoffset())->GetChainStates();
                                 ^
libBAT_rdict.cxx:6931:33: error: template argument 2 is invalid
libBAT_rdict.cxx:6931:40: error: invalid type in declaration before '=' token
          const vector<ChainState>& obj = ((const BCEngineMCMC*) G__getstructoffset())->GetChainStates();
                                        ^
libBAT_rdict.cxx:6931:103: error: invalid initialization of reference of type 'const int&' from expression of type 'const std::vector<BCEngineMCMC::ChainState>'
          const vector<ChainState>& obj = ((const BCEngineMCMC*) G__getstructoffset())->GetChainStates();
```

fredRos commented on BAT/BCEngineMCMC.h:
```diff
@@ -272,34 +289,43 @@ class BCEngineMCMC
+    /**
+     * @param c index of the Markov chain
+     * @return state of chains */
+    const ChainState& GetChainStates(unsigned c) const
```
If we have `GetChainStates()`, we don't need `GetChainStates(unsigned c)` in my opinion

fredRos replied:
Actually the opposite is better. Let's remove `GetChainStates()` and keep `GetChainState(unsigned c)`. Note the missing `s` on `...State`. This gives us the opportunity to change the underlying container if we wanted to.

greenwald replied:
I was just making it match to `Getx()` and `GetProb()`. I don't think we should have those either now. But for back-compatibility I left them in.

fredRos commented on BAT/BCEngineMCMC.h:
```diff
@@ -272,34 +289,43 @@ class BCEngineMCMC
      * @return current point of each Markov chain */
-    const std::vector<std::vector<double> >& Getx() const
-    { return fMCMCx; }
+    std::vector<std::vector<double> > Getx() const;
```
Carefully check where Getx() is called? Changing from reference to value leads to calls like this producing undefined behavior

```c++
std::vector<std::vector<double> > v(Getx().begin(), Getx().end();
```

If we can remove it because we have `GetX(unsigned c)`, we should

greenwald replied:
`Getx()` is never called inside any of our code. I'd be happy to remove it.

fredRos commented on BAT/BCEngineMCMC.h:
```diff
@@ -1411,13 +1450,29 @@ class BCEngineMCMC
 private:
+
+    // grant friendship to BCModel to access the ThreadLocalStorage
+    friend class BCModel;
```
Shouldn't `ThreadLocalStorage` then be in the `protected` section?

greenwald replied:
I assumed you put it in private so that no _user_ classes inheriting from `BCEngineMCMC` could touch it (which makes sense for preserving thread safety). If we made it `protected` couldn't any class derived from `BCModel` (i.e. all user models) access it? As it is now, unless I'm wrong, `BCModel` can access it but classes derived from `BCModel` cannot.

fredRos commented on BAT/BCEngineMCMC.h:
```diff
@@ -1411,13 +1450,29 @@ class BCEngineMCMC
+
+        /**
+         * Store log_prior of local proposal point */
+        double log_prior;
```
I don't like this coupling with `BCModel` and the prior. We worked hard to separate things a bit and keep the prior out of the engine

fredRos commented on src/BCModel.cxx:
```diff
@@ -87,16 +87,17 @@ BCModel::~BCModel()
+        fMCMCThreadLocalStorage[chain].log_likelihood = -std::numeric_limits<double>::infinity();
 
-    return ll + lp;
+    return fMCMCThreadLocalStorage[chain].log_probability = fMCMCThreadLocalStorage[chain].log_prior + fMCMCThreadLocalStorage[chain].log_likelihood;
```
I don't like assignment within the return statement.

fredRos commented on src/BCEngineMCMC.cxx:
```diff
@@ -168,13 +167,9 @@ BCEngineMCMC::BCEngineMCMC(const BCEngineMCMC& other)
       fMCMCPhase(other.fMCMCPhase),
-      fMCMCx(other.fMCMCx),
-      fMCMCObservables(other.fMCMCObservables),
+      fMCMCStates(other.fMCMCStates),
```
Nice to see the number of members reduced

fredRos commented on src/BCEngineMCMC.cxx:
```diff
@@ -1239,40 +1230,40 @@ void BCEngineMCMC::Remarginalize(bool autorange)
+        fMCMCTree_State.iteration = 0;
 
     // link log(likelihood) if available
     if (fMCMCTree->GetBranch("LogLikelihood"))
```
Why is and was likelihood used here? I thought we had [`BCModel::InitializeMarkovChainTree` for that. We shouldn't have any variable named  `likelihood` inside the engine

greenwald replied:
I don't understand your comment in reference to the highlighted code.

This function is meant to load in a `TTree` and fill all possible marginalization information given the information in the tree. It's structured to not presume the tree has been made by BAT. At it's minimum, I believe it only requires parameters be in the tree.

fredRos commented on src/BCEngineMCMC.cxx:
```diff
@@ -1301,27 +1292,21 @@ void BCEngineMCMC::Remarginalize(bool autorange)
-                        if (fMCMCTree_Observables[i] > XMax[fMCMCTree_Parameters.size() + i])
-                            XMax[fMCMCTree_Parameters.size() + i] = fMCMCTree_Observables[i];
+                    for (unsigned i = 0; i < fMCMCTree_State.observables.size(); ++i) {
+                        XMin[GetNParameters() + i] = std::min(XMin[GetNParameters() + i], fMCMCTree_State.observables[i]);
```
congrats on simplifying the code

fredRos commented on src/BCEngineMCMC.cxx:
```diff
@@ -1519,90 +1507,77 @@ bool BCEngineMCMC::GetProposalPointMetropolis(unsigned ichain, unsigned ipar, st
 }
 
+// --------------------------------------------------------
+bool BCEngineMCMC::AcceptOrRejectPoint(unsigned chain, unsigned parameter)
```
Good job: promoting code reuse

fredRos commented on src/BCEngineMCMC.cxx:
```diff
@@ -1631,7 +1606,7 @@ bool BCEngineMCMC::GetNewPointMetropolis()
             for (unsigned ichain = 0; ichain < fMCMCNChains; ++ichain) {
                 UpdateChainIndex(ichain);
-                return_value *= GetNewPointMetropolis(ichain, ipar);
+                return_value &= GetNewPointMetropolis(ichain, ipar);
```
the bitwise operator `&=` seems appropriate here because it does not short-circuit and we want the rhs to be evaluated.

greenwald replied:
Appropriate or inappropriate? I'm confused by your comment. You are approving of the change or disapproving?

fredRos commented on src/BCEngineMCMC.cxx:
```diff
@@ -1657,41 +1632,62 @@ bool BCEngineMCMC::GetNewPointMetropolis()
+    ////////////////////////////////////////
+    // fill each 1-dimensional histogram that exists
+    for (unsigned i = 0; i < GetNVariables() && i < fH1Marginalized.size(); ++i)
+        if (fH1Marginalized[i]) {
```
Or more concisely

```c++
if (TH1* h = fH1Marginalized[i]) {
...
h->Fill(...)
```

fredRos commented on src/BCEngineMCMC.cxx:
```diff
@@ -2307,24 +2300,16 @@ void BCEngineMCMC::MCMCInitialize()
-    fMCMCLogLikelihood.assign(fMCMCNChains, -std::numeric_limits<double>::infinity());
-    fMCMCLogLikelihood_Provisional.assign(fMCMCNChains, -std::numeric_limits<double>::infinity());
+    // reset states holder
+    fMCMCStates.assign(fMCMCNChains, ChainState(GetNObservables()));
```
Why `GetNObservables()`. How are the number of parameters passed to `ChainState`?

greenwald replied:
By assignment from `ThreadLocalStorage`. Only the `observables` vector needs to be given a size, since it is assigned into directly from a loop over the `ObservablesSet`.

fredRos commented on src/BCEngineMCMC.cxx:
```diff
@@ -2392,106 +2374,108 @@ void BCEngineMCMC::MCMCInitialize()
-            break;
+    // keep previous values
+    case kInitContinue : {
+        throw std::runtime_error("BCEngineMCMC::MCMCInitialize : Continuing chains not yet supported. Sorry!");
```
Didn't @tpoeschl have a patch for continuing?

greenwald replied:
He's handed me his code. I will see how easy it is to integrate. I think it mostly involves adding some getters and setters.

fredRos commented on src/BCEngineMCMC.cxx:
```diff
@@ -3745,6 +3729,33 @@ void BCEngineMCMC::Statistics::Reset(bool reset_mode, bool reset_efficiency)
+// ---------------------------------------------------------
+std::vector<std::vector<double> > BCEngineMCMC::Getx() const
+{
+    std::vector<std::vector<double> > x;
```
I think it's nicer to set `x` to the known size instead of using `push_back`

greenwald replied:
I've removed this code.

fredRos commented:
This is hard to debug but I found some interesting properties by the printf approach. Here is what I updated:

```c++
void BCEngineMCMC::InChainFillTree(const ChainState& cs, unsigned chain_number)
{
    if (!fMCMCTree)
        return;
    fMCMCTree_Chain = chain_number;
    fMCMCTree_State = cs;
    if (cs.log_probability != (cs.log_prior + cs.log_likelihood)) {
        BCLog::OutError(Form("mismatch for chain %u in iteration %d: state (%g != %g, %g), threadlocal (%g =? %g, %g)", chain_number, fMCMCCurrentIteration,
                         cs.log_probability, cs.log_likelihood, cs.log_prior,
                         fMCMCThreadLocalStorage[chain_number].log_probability, fMCMCThreadLocalStorage[chain_number].log_likelihood, fMCMCThreadLocalStorage[chain_number].log_prior
                        ));
    }
    fMCMCTree->Fill();
}
```

Always using 8 chains in the above MWE, I see no mismatches for 1 or 2 threads. With three threads, I only get mismatches in chain 1. With 4 threads, up to chain 3. With 8 threads, I see mismatches in all chains, so including chain 7. The pattern is the same in all cases: threadlocal has no mismatch, only the chain state has a mismatch between logprop and loglikelihood. The mismatches continue for some number of iterations, I guess until a proposed point is accepted.

```
Error   : mismatch for chain 1 in iteration 25006: state (-0.80768 != -1.05322, 0), threadlocal (-1.05322 =? -1.05322, 0)
Error   : mismatch for chain 1 in iteration 25007: state (-0.80768 != -1.05322, 0), threadlocal (-1.05322 =? -1.05322, 0)
Error   : mismatch for chain 1 in iteration 25008: state (-0.80768 != -1.05322, 0), threadlocal (-1.05322 =? -1.05322, 0)
```

I feel that the copying to the `state.log_probability` is somehow messed up

greenwald commented:
Concerning our (offline) discussion of `UpdateChainIndex`: I think we still need to call this function before calling anything leading to `LogEval` being called. Even when not running parallel, the chain number should be something accessible to the user in the likelihood/prior call (or `CalculateObservabldes`), and this will not be the case unless `UpdateChainIndex` is called.

greenwald commented:
I have implemented all @fredRos's requested changes. And simplified the code a little by implementing an assignment operator for `ChainState = ThreadLocalStorage` (essentially performing proposal acceptance).

The initialization problem is solved.

I can resolve the timing issue, by removing these lines from `AcceptOrRejectPoint`
```
// this line shouldn't be necessary
fMCMCStates[chain].log_probability = p1;
```
However, I then get an error in `BCModel_TEST`:
> Test case failed: 
> BCModel_TEST.cxx:82:One relative error of 'm.GetMarginalizedHistogram(2)->GetRMS()' = '1.05921900210098' and 'm.sigma()' = '1' is greater than 0.05. The results are 0.05590817573 and 0.0592190021
> FAIL BCModel.TEST (exit status: 1)

So it's just outside the range you've allowed in the test (5%). This could just failing the arbitrary 5% choice, and not an actual problem. You decide.

fredRos commented:
Sorry, github doesn't let me answer inline to the comments because it says it can't find the code anymore:

> Appropriate or inappropriate? I'm confused by your comment. You are approving of the change or disapproving?

I'm approving that one

> I don't understand your comment in reference to the highlighted code.

we discussed this offline. I thought the term `likelihood`shouldn't even appear in the engine but we give up such high principles for BAT 1.0

greenwald commented:
So if I understand correct, all changes are approved in this pull request so far. But we have yet to diagnose or solve the actual race condition issue.

fredRos commented:
Yes, the changes I see are ok. But the issue isn't solved. I'm still scratching my head but I found out a bit more. In the log output below, chain 0 and chain 3 appear mixed up at the same time. And apparently  the chain index `0` appears twice! The other mismatches then must be consequences of reading and writing to chain 0 or 3 in parallel. 
I changed

```c++
bool BCEngineMCMC::AcceptOrRejectPoint(unsigned chain, unsigned parameter)
{
    // retrieve current probability
    double p0 = fMCMCStates[chain].log_probability;
    if (!std::isfinite(p0)) p0 = -std::numeric_limits<double>::max();
    // calculate proposed probability
    const double p1 = LogEval(fMCMCThreadLocalStorage[chain].parameters);

    // if the new point is more probable, keep it; or else throw dice
    if (std::isfinite(p1) && (p1 >= p0 || log(fMCMCThreadLocalStorage[chain].rng->Rndm()) < (p1 - p0))) {
        // accept point
        fMCMCStates[chain] = fMCMCThreadLocalStorage[chain];
        if (fMCMCStates[chain].log_probability != p1) {
            BCLog::OutError(Form("mismatch for chain %u in iteration %d: state logprob = %g, p1 = %g", chain, fMCMCCurrentIteration,
                                 fMCMCStates[chain].log_probability, p1
                                ));
        }
        // this line shouldn't be necessary
        fMCMCStates[chain].log_probability = p1;
```

```
Detail  :  --> iteration number   8000 ( 80 %)
Error   :  --> iteramismatch for chain 0 in iteration 8007: state logprob = -24.4247, p1 = -24.4298

Error   :  --> iteramismatch for chain 0 in iteration 8007: state logprob = -24.4247, p1 = -24.4298

Error   : mismatch for chain 0 in iteration 8008: state (-24.4298 != -7.41875, -17.006), threadlocal (-24.4247 =? -7.41875, -17.006)
Error   : mismatch for chain 3 in iteration 8008: state (-24.4247 != -6.56026, -17.006), threadlocal (-23.5662 =? -6.56026, -17.006)
Error   : mismatch for chain 0 in iteration 8009: state (-24.4298 != -7.41875, -17.006), threadlocal (-28.4131 =? -11.4071, -17.006)
Error   : mismatch for chain 3 in iteration 8009: state (-24.4247 != -6.56026, -17.006), threadlocal (-27.4665 =? -10.4605, -17.006)
Error   : mismatch for chain 0 in iteration 8010: state (-24.4298 != -7.41875, -17.006), threadlocal (-28.0188 =? -11.0128, -17.006)
Error   : mismatch for chain 0 in iteration 8011: state (-24.4298 != -7.41875, -17.006), threadlocal (-26.1857 =? -9.17968, -17.006)
Detail  :  --> iteration number   9000 ( 90 %)
```

To investigate this, I wondered if `private(ichain)` in `#pragma omp parallel for` was a problem, removed it, and still get errors.

fredRos commented:
From the above, it seemed clear that `GetCurrentChain()` doesn't always work correctly. Perhaps it's because of what inside `UpdateChainIndex` or just where its' called. I can't figure out. But I just tried passing on the chain index explicity through the function. So I have

```c++
    const double p1 = LogEval(fMCMCThreadLocalStorage[chain].parameters, chain);
```

and duplicated the existing `BCModel::LogProbabilityNN` w/o calling `GetCurrentChain()`:

```c++
double BCModel::LogEval(const std::vector<double>& parameters, unsigned chain)
{
    ThreadLocalStorage& s = fMCMCThreadLocalStorage[chain];
    // first calculate prior (which is usually cheaper than likelihood)
    s.log_prior = LogAPrioriProbability(parameters);
    // then calculation likelihood, or set to -inf, if prior already invalid
    if (std::isfinite(s.log_prior))
        s.log_likelihood = LogLikelihood(parameters);
    else
        s.log_likelihood = -std::numeric_limits<double>::infinity();

    s.log_probability = s.log_prior + s.log_likelihood;
    return s.log_probability;
}
```

Then `parallel_TEST` passes, and I repeated 10 times. This is no guarantee that the problem is fixed but I feel we have a workaround now.

fredRos commented:
My last comment doesn't fix `GetCurrentChain`. User classes that rely on it may still suffer from the bug. 

For (my) future reference: problems in multithreaded programs usually (always?) come from race conditions; i.e. concurrent read/write. To solve the issue, find the memory or variable that is written while read. In our case, `GetCurrentChain` occasionally fails. Inside there is a loop over a `std::map` in 

```c++
    ChainIndex_t::const_iterator it = fChainIndex.find(threadIndex);
    if (it == fChainIndex.end())
        return defaultChain;
    return it->second;
```

In `BCEngineMCMC::GetNewPointMetropolis()`, the `fChainIndex is unnecessarily cleared in every iteration. So 

```c++
fChainIndex.clear()
...
            #pragma omp parallel for shared(chunk) schedule(static, chunk)
            for (unsigned ichain = 0; ichain < fMCMCNChains; ++ichain) {
                UpdateChainIndex(ichain);
                return_value &= GetNewPointMetropolis(ichain, ipar);
}
```
so `UpdateChainIndex(ichain)` adds a new element to the map while in another thread `GetCurrentChain` is called inside `LogEval` and it happens to get an invalid iterator if the map is changed after `find` and before `return it->second;`. This is unlikely and so the error happens only seldomly.

The fix: I commented `fChainIndex.clear();` in `GetNewPointMetropolis` and now I don't get the failed test. But the race condition may still occur in the first iteration. So I removed all such `.clear()` calls and inserted elements into the map inside `SyncThreadStorage` in commit dcd3657

Please add the extra commits in my fork of the `user_init` branch, then I can approve  the PR

fredRos commented:
I now replaced `std::map` with `std::vector`. It is (unnoticeably) faster, reduces code size, and would have avoided the race condition. Please cherry pick my latest commits so I can approve and merge this PR
